driver/src/orderbook/retrying_retrieval.rs

```diff
@@ -0,0 +1,222 @@
+        block = retry.select_next_some() => block,
+    };
+
+    let mut reader = PaginatedAuctionDataReader::new(batch_id.into());
+    loop {
+        let mut page_future = contract
+            .encoded_orders_paginated(
+                reader.pagination().previous_page_user,
+                reader
+                    .pagination()
+                    .previous_page_user_offset
+                    .try_into()
+                    .expect("user cannot have more than u16::MAX orders"),
+                page_size,
+                block_number.into(),
+            )
+            .fuse();
+
+        let page = select! {
+            page = page_future => page?,
+            block = retry.next() => {
+                block_number = match block {
+                    Some(block) => block,
+                    None => {
+                        return Err(anyhow!("retry channel unexpectedly closed"));
+                    }
+                };
+                reader = PaginatedAuctionDataReader::new(batch_id.into());
+                continue;
+            },
+        };
+
+        let number_of_orders: u16 = reader
+            .apply_page(&page)
+            .try_into()
+            .expect("number of orders per page should never overflow a u16");
+
+        if number_of_orders < page_size {
+            return Ok(reader.get_auction_data());
+        }
+    }
+}
+
+/// Watches exchange contract events for events that would invalidate an
+/// orderbook.
+///
+/// This event uses a `retry` channel to signal the most accurate block number
+/// to use to re-query the orderbook in case an event is encoutered that would
+/// otherwise invalidate the orderbook
+///
+/// Note this takes an additional `done` channel that can be used to request
+/// this future to end. If the done channel is never signaled, then the future
+/// returned by this method will not resolve unless it encounters an error.
+async fn watch_contract_events(
+    contract: &dyn StableXContractAsync,
+    batch_id: u32,
+    mut retry: mpsc::Sender<u64>,
+    done: oneshot::Receiver<()>,
+) -> Result<()> {
+    use batch_exchange::Event::*;
+    use EventData::*;
+
+    let mut done = done.fuse();
+    let mut events = contract.all_events().fuse();
+
+    loop {
+        // NOTE: Wait for the next event. Simultaneously poll the done channel
+        //   to check to see if the caller requested us to exit.
+        let event = select! {
+            event = events.next() => match event.transpose()? {
+                Some(event) => event,
+                None => break Err(anyhow!("event stream unexpectedly ended")),
+            },
+            _ = done => break Ok(()),
+        };
+
+        let block_number = match event.meta.as_ref() {
+            Some(meta) => meta.block_number,
+            _ => return Err(anyhow!("unexpected event on pending block in event stream")),
+        };
+
+        let requires_retry = match &event.data {
```

Comment:
Could we simplify the logic, by checking if the event's timestamp is from the next batch or the one we are currently trying to solve. In the former case we know that `requires_retry` will be false regardless of the event type, wouldn't we? We then would only have to look at the event details for order placements.

Reply:
Sounds like a good plan to me.